### ISSUE TITLE
Reader.run() should throttle handshake attempts

### ISSUE THREAD

**Issue by abuchanan-nr** [segmentio/kafka-go#284]
**Describe the bug**
I've seen a message like the following printed dozens of times a second:
```
rebalance failed for consumer group, {group name}: joinGroup failed: [16] Not Coordinator For Group: the broker returns this error code if it receives an offset fetch or commit request for a group that it is not a coordinator for
```

**Kafka Version**
0.10

**Additional context**
I suspect this is because of this for-loop: https://github.com/segmentio/kafka-go/blob/master/reader.go#L881

Maybe there should be some throttling on that?

**Comment by stevevls** [comment]
I was just looking at this code the other day and thinking it might be susceptible to this type of error.  I believe it retries immediately under the assumption that it got a rebalance error and that the join group call will then block up to the rebalance timeout.

Will make sure to fix this in #277

**Comment by stevevls** [comment]
Heh heh...just went to drop a todo in the new consumer group code and found that I had left this comment:
> // we're not currently sleeping or doing anything to throttle
> // the loop in case of an error.  this is the behavior of the
> // original consumer group impl.  this likely could be improved.

**Comment by stevevls** [comment]
Just merged #277, which adds a `JoinGroupBackoff` parameter that will introduce a sleep time when an unexpected error occurs while joining and syncing the group.  Default value is `5s`.
